docs/syz_verifier.md

```diff
@@ -0,0 +1,81 @@
+# syz-verifier
+
+Logical bugs are a type of  bugs that make a program operate incorrectly, possibly
+producing unintended output, but usually without causing the program to crash or
+producing assertion failures. Since programs with logical bugs generally compile,
+these are hard to detect and require more than dynamic analysis tools.
+
+Differential fuzzing aims to detect such bugs by providing the same input to different
+implementations (whose specifications say they behave identically) and cross-compares
+their resulting behaviour to determine whether its the same.
+
+`syz-verifier` is a differential fuzzing tool that cross-compares the execution of programs,
+generated using `syzlang`, on different versions of the Linux kernel.
+
+The architecture of `syz-verifier` is shown in the following diagram.
+
+
+![Architecture overview](syz_verifier_structure.png)
+
+The `syz-verifier` process wiil start and manage VM instances created from the kernels to be cross-compared. It will also start the `syz-runner` process on the VMs. Communication between the host and the guest will be done via RPC calls.
+
+`syz-verifier` will generate and send a continuous stream of programa to `syz-runner` via RPC  while `syz-runner` is responsible for starting `syz-executor` processes and turning the program into input for those. `syz-executor` will process the input, which triggers a sequence of syscalls in the kernel. Then, `syz-runner` collects the results and sends them back to the host.
+
+ When `syz-verifier` has received results from all the kernels for a specific program, it verifies them to ensure they are identical. If a mismatch is found, the program will be rerun on all the kernels to ensure the mismatch is not flaky (i.e. it didn't occur because of some background activity or external state). If the mismatch occurs in all reruns, `syz-verifier` will create a report for the program and write it to persistent storage.
+
+
+# How to use `syz-verifier`
+
+The tool is not currently integrated with `syzbot` so it needs to be run manually.
+
+After cloning the repository (see how [here](/docs/linux/setup.md#go-and-syzkaller)), build the tool as:
+
+```
+make verifier runner
+```
+
+The resulting executable can then be found in `$SYZ/bin`, where `$SYZ` is the location where the `syzkaller` project was cloned,
+
+To start using the tool, separate configuration files need to be created for each kernel you want to include in the verification. An example of Linux configs can be found [here](/docs/linux/setup_ubuntu-host_qemu-vm_x86-64-kernel.md#syzkaller).
```

Comment:
We need to somehow say that these configuration files are the same configuration files used for syz-manager.